src/main/java/uk/ac/sanger/sccp/stan/service/history/HistoryServiceImp.java

```diff
@@ -78,11 +80,131 @@ public History getHistoryForSampleId(int sampleId) {
+        Collection<Operation> ops = opIds.isEmpty() ? List.of() : BasicUtils.asCollection(opRepo.findAllById(opIds));
+        List<Release> releases = releaseIds.isEmpty() ? List.of() : releaseRepo.findAllByIdIn(releaseIds);
+        Set<Integer> labwareIds = labwareIdsFromOps(ops);
+        List<Labware> opLabware = lwRepo.findAllByIdIn(labwareIds);
+        //If barcode given, filter the releases and opLabware by barcode
+        if(barcode!=null) {
+            releases = releases.stream().filter(release->release.getLabware().getBarcode().equalsIgnoreCase(barcode)).collect(Collectors.toList());
+            opLabware = opLabware.stream().filter(lw->lw.getBarcode().equals(barcode)).collect(Collectors.toList());
+        }
+        //If externalName or donorName given, filter the samples by externalName or donorName
+        Set<Integer> sampleIds = null;
+        if(externalName!=null || donorName!=null) {
+            List<Sample> filteredSamples = getFilteredSamples(externalName, donorName, null);
+            if (filteredSamples != null) {
+                sampleIds = filteredSamples.stream().map(Sample::getId).collect(toSet());
+            }
+        }
+        List<HistoryEntry> opEntries = createEntriesForOps(ops, sampleIds, opLabware, null, work.getWorkNumber());
+        final List<HistoryEntry> releaseEntries = createEntriesForReleases(releases, sampleIds, null, work.getWorkNumber());
+        List<HistoryEntry> entries = BasicUtils.concat(opEntries, releaseEntries);
+        List<Labware> allLabware = new ArrayList<>(opLabware);
+        if (!releases.isEmpty()) {
+            labwareIds = new HashSet<>(labwareIds);
+            for (Release rel : releases) {
+                Labware lw = rel.getLabware();
+                if (labwareIds.add(lw.getId())) {
+                    allLabware.add(lw);
+                }
+            }
+        }
+        List<Sample> samples = referencedSamples(entries, allLabware);
+        entries.sort(Comparator.comparing(HistoryEntry::getTime));
+        List<SamplePositionResult> samplePositionResults = slotRegionService.loadSamplePositionResultsForLabware(allLabware);
+        return new History(entries, samples, allLabware, samplePositionResults);
+    }
+
+
     @Override
     public History getHistoryForExternalName(String externalName) {
         List<Tissue> tissues;
-        if (externalName!=null && externalName.indexOf('*') >= 0) {
-            tissues = tissueRepo.findAllByExternalNameLike(BasicUtils.wildcardToLikeSql(externalName));
+        // If the externalName contains a wildcard, check if it is a wildcard or * contained in the externalName.
+       if (externalName!=null && externalName.indexOf('*') >= 0) {
+            // Create a regex pattern from the wildcard pattern.
+            Pattern pattern = BasicUtils.makeWildcardPattern(externalName);
+            // Use the pattern to create a matcher for the externalName.
+            Matcher matcher = pattern.matcher(externalName);
```

Comment:
It looks like you're converting `externalName` to a regular expression, and then testing if it matches itself. That doesn't make sense. If you have some tissues loaded already, you can check their external names against the regular expression; and if you haven't, you don't need the regular expression.